### PR TITLE
Add an option to anonymize users

### PR DESCRIPTION
**Pull request by tomasr8**
Adds a CLI command which allows Indico admins to anonymize users for the purposes of GDPR/OC11.
We only delete strictly user data, anything connected to events - data in person links, registrations, etc.. is the responsibility of event managers.

### CODE REVIEW

File: indico/cli/user.py [hunk 1]
```diff
@@ -217,6 +219,42 @@ def unblock(user_id):
...
+    This will remove all personal data about the given user including
+    their name, affiliation and emails. Note that any data tied to events
+    such registrations will not be deleted.
+    """
+    user = User.get(user_id)
+    if user is None:
+        click.secho('This user does not exist', fg='red')
+        return
+    _print_user_info(user)
+    if click.confirm(click.style('Anonymize this user?', fg='yellow')):
```
Comment (ThiefMaster) [inline indico/cli/user.py:236]: `if not ...: return`

and I'd move all the logic to a util in indico/module/users

File: indico/cli/user.py [hunk 2]
```diff
@@ -217,6 +219,42 @@ def unblock(user_id):
...
+        user.phone = ''
+        user.address = ''
+        user.picture = None
+        user.picture_metadata = None
+        user.picture_source = ProfilePictureSource.standard
+        user.favorite_users = set()
+        user.favorite_events = set()
+        user.favorite_categories = set()
+        user.identities = set()
+        # TODO: check what breaks when we delete all emails
```
Comment (ThiefMaster) [inline indico/cli/user.py:251]: I think the system user also has no email so you can check for places where we check if a user is the system user and see if there's anything important

File: indico/cli/user.py [hunk 3]
```diff
@@ -217,6 +219,42 @@ def unblock(user_id):
...
+        UserEmail.query.filter(UserEmail.user == user).delete()
+        # TODO: More things we can delete?
```
Comment (ThiefMaster) [inline indico/cli/user.py:253]: I'd soft-delete the user itself

**Comment by ThiefMaster** [comment]
We could also add an `indico user delete` cli which attempts to hard-delete the user. This would of course fail if the ussr is linked to things that do not support nulling the foreign key, but e.g. for spam users that didn't do anything in indico or users who just registered for an event this would work nicely.

File: indico/modules/events/export.py [hunk 4]
```diff
@@ -262,6 +262,7 @@ def _make_idref(self, column, value, incoming=False, target_column=None):
         uuid = self.id_map[fullname].setdefault(value, self._get_uuid())
         if type_ == 'userref' and uuid not in self.users:
             user = User.get(value)
+            # TODO: Come back to this one
```
Comment (tomasr8) [inline indico/modules/events/export.py:265]: This is not needed anymore ;)

File: indico/modules/users/util.py [hunk 5]
```diff
@@ -409,6 +411,60 @@ def merge_users(source, target, force=False):
...
+    user.address = ''
+    user.picture = None
+    user.picture_metadata = None
+    user.picture_source = ProfilePictureSource.standard
+    user.favorite_users = set()
+    user.favorite_events = set()
+    user.favorite_categories = set()
+    user.identities = set()
+
+    # TODO: check what breaks when we delete all emails
```
Comment (tomasr8) [inline indico/modules/users/util.py:437]: neither this

File: indico/modules/users/util.py [hunk 6]
```diff
@@ -409,6 +411,60 @@ def merge_users(source, target, force=False):
...
+        EventPrincipal, CategoryPrincipal, AttachmentPrincipal, AttachmentFolderPrincipal,
+        BlockingPrincipal, TrackPrincipal, EventSettingPrincipal, SettingPrincipal, RoomPrincipal,
+        SessionPrincipal
+    ]
+
+    for principal in principals:
+        principal.query.filter(principal.user == user).delete()
+
+    user.is_deleted = True
+    # TODO: More things we can delete?
```
Comment (tomasr8) [inline indico/modules/users/util.py:464]: I think we already covered everything

File: indico/modules/users/util.py [hunk 7]
```diff
@@ -419,6 +421,57 @@ def merge_users(source, target, force=False):
...
+
+    categories = user.category_roles.copy()
+    for category in categories:
+        category.members.remove(user)
+
+    principals = [
+        EventPrincipal, CategoryPrincipal, AttachmentPrincipal, AttachmentFolderPrincipal,
+        BlockingPrincipal, TrackPrincipal, EventSettingPrincipal, SettingPrincipal, RoomPrincipal,
+        SessionPrincipal
+    ]
```
Comment (ThiefMaster) [inline indico/modules/users/util.py:467]: This should not be hardcoded. You can use this to get a list of all principal classes:

```py
    classes = [sc for sc in [*PrincipalMixin.__subclasses__(), *PrincipalPermissionsMixin.__subclasses__()]
               if hasattr(sc, 'query')]
```

**Review by ThiefMaster — APPROVED** [review]

**Pull request merged**